File: rust/vectors/composites.go
package main

// The composite corpus: compact-mode messages carrying nested structs, arrays of
// structs, maps and nested slices.
//
// It is a second file rather than more cases in vectors.json because the schema
// representation differs. A flat schema names each field's kind with a string
// ("i32", "strs"); a composite one needs a tree, since a nested struct carries a
// whole field list of its own. Keeping them apart leaves the original corpus and
// its parser untouched, which is worth more than one uniform file: those 26
// messages pin the format as it was before composites existed.
//
// As in main.go, every expected value is read back out of the message by the
// real Go compact reader rather than derived from the input, so the corpus
// records what the format actually wrote -- presence included, which for compact
// mode is the whole of the omit-zero rule.

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"

	"github.com/ivanjoz/colbin"
	"github.com/ivanjoz/colbin/compact"
)

// --- the types ---------------------------------------------------------------

// ccInner is tagged so its ids fit the narrow key, which is what lets a nested
// message use 4-bit keys at all: the key width is message-wide, so every
// reachable struct has to be tagged.
type ccInner struct {
	X int32  `cb:"1"`
	Y string `cb:"2"`
	Z bool   `cb:"3"`
}

// ccUntagged has a hashed id, so any message reaching it pays 8-bit keys.
type ccUntagged struct {
	A int32
}

type ccNested struct {
	ID   int64   `cb:"1"`
	Sub  ccInner `cb:"2"`
	Name string  `cb:"3"`
}

type ccWideNested struct {
	Sub ccUntagged `cb:"1"`
}

type ccRows struct {
	ID   int64     `cb:"1"`
	Rows []ccInner `cb:"2"`
}

type ccMaps struct {
	S map[string]int32   `cb:"1"`
	I map[int32]string   `cb:"2"`
	V map[string]ccInner `cb:"3"`
	L map[string][]int32 `cb:"4"`
}

type ccNestedSlices struct {
	II [][]int32          `cb:"1"`
	SS [][]string         `cb:"2"`
	BB [][]byte           `cb:"3"`
	MM []map[string]int32 `cb:"4"`
}

// ccDeep is three struct levels plus a map of arrays of structs, so the
// recursion is exercised past the one level a single nested field proves.
type ccDeep struct {
	ID    int32                `cb:"1"`
	Mid   ccNested             `cb:"2"`
	Deep  map[string][]ccInner `cb:"3"`
	Lists [][]ccInner          `cb:"4"`
}

// --- output shapes -----------------------------------------------------------

// kindNode is a kind that may hold kinds. The scalar forms keep the same names
// main.go uses, so the two corpora agree on everything they share.
type kindNode struct {
	Kind   string      `json:"kind"`
	Fields []schemaOut `json:"fields,omitempty"` // kind == "struct"
	Elem   *kindNode   `json:"elem,omitempty"`   // kind == "array"
	Key    *kindNode   `json:"key,omitempty"`    // kind == "map"
	Val    *kindNode   `json:"val,omitempty"`    // kind == "map"
}

type schemaOut struct {
	Name string   `json:"name"`
	ID   int      `json:"id"`
	Kind kindNode `json:"k"`
}

type ccCaseOut struct {
	Name        string      `json:"name"`
	Shape       int         `json:"shape"`
	AllPositive bool        `json:"allPositive"`
	NarrowKeys  bool        `json:"narrowKeys"`
	Fields      []schemaOut `json:"fields"`
	Message     string      `json:"message"`
	Records     [][]ccField `json:"records"`
	// See caseOut.RustByteExact in main.go.
	RustByteExact bool `json:"rustByteExact"`
}

type ccField struct {
	ID int `json:"id"`
	V  any `json:"v"`
}

type ccEntry struct {
	K any `json:"k"`
	V any `json:"v"`
}

type ccCorpus struct {
	Cases []ccCaseOut `json:"cases"`
}

// writeComposites generates the corpus. Every case is forced through compact
// mode: Marshal picks the mode on size for a composite type, and a corpus case
// that silently came out columnar would test nothing here.
func writeComposites(out string) {
	var cases []ccCaseOut
	add := func(name string, value any) {
		cases = append(cases, compositeCase(name, value))
	}

	add("cc-nested-struct", ccNested{ID: 7, Sub: ccInner{X: -3, Y: "inner", Z: true}, Name: "outer"})
	// The nested struct holds nothing, so it is omitted entirely and the record
	// never names field 2.
	add("cc-nested-zero", ccNested{ID: 7, Name: "outer"})
	add("cc-nested-only", ccNested{Sub: ccInner{Y: "just the nested one"}})
	add("cc-nested-wide-keys", ccWideNested{Sub: ccUntagged{A: 5}})

	add("cc-rows", ccRows{ID: 1, Rows: []ccInner{{X: 1, Y: "a"}, {X: -2, Y: "b", Z: true}, {}}})
	add("cc-rows-one", ccRows{ID: 1, Rows: []ccInner{{X: 9}}})
	add("cc-rows-empty", ccRows{ID: 1})

	add("cc-maps", ccMaps{
		// One entry per map: Go's iteration order is unspecified, so a corpus
		// case with two would pin an order the encoder does not promise.
		S: map[string]int32{"alpha": -2},
		I: map[int32]string{-7: "neg"},
		V: map[string]ccInner{"x": {X: 1, Y: "y", Z: true}},
		L: map[string][]int32{"nums": {1, 2, 3}},
	})
	add("cc-maps-empty", ccMaps{S: map[string]int32{"only": 1}})

	add("cc-nested-slices", ccNestedSlices{
		II: [][]int32{{1, 2, 3}, nil, {-4}},
		SS: [][]string{{"a", ""}},
		BB: [][]byte{{1, 2}, {}},
		MM: []map[string]int32{{"k": 1}},
	})

	add("cc-deep", ccDeep{
		ID:    -12345,
		Mid:   ccNested{ID: 1, Sub: ccInner{X: 42, Y: "mid", Z: true}, Name: "middle"},
		Deep:  map[string][]ccInner{"k": {{X: 1, Y: "deep"}, {}}},
		Lists: [][]ccInner{{{X: 7}}, nil},
	})

	// Two and three records, so the array shapes carry composites too.
	add("cc-shape-2", []ccNested{
		{ID: 1, Sub: ccInner{X: 1}},
		{ID: 2, Name: "two"},
	})
	add("cc-shape-3", []ccRows{
		{ID: 1, Rows: []ccInner{{X: 1}}},
		{ID: 2},
		{ID: 3, Rows: []ccInner{{Y: "three"}, {Z: true}}},
	})

	body, err := json.MarshalIndent(ccCorpus{Cases: cases}, "", "  ")
	if err != nil {
		panic(err)
	}
	if err := os.MkdirAll(filepath.Dir(out), 0o755); err != nil {
		panic(err)
	}
	if err := os.WriteFile(out, append(body, '\n'), 0o644); err != nil {
		panic(err)
	}
	fmt.Printf("%d composite messages -> %s\n", len(cases), out)
}

func compositeCase(name string, value any) ccCaseOut {
	data, err := colbin.MarshalForceCompact(value)
	if err != nil {
		panic(fmt.Sprintf("%s: %v", name, err))
	}
	elem := reflect.TypeOf(value)
	if elem.Kind() == reflect.Slice {
		elem = elem.Elem()
	}
	fields := compositeSchema(elem)

	r, err := compact.NewReader(data)
	if err != nil {
		panic(fmt.Sprintf("%s: %v", name, err))
	}
	c := ccCaseOut{
		Name:        name,
		Shape:       int(r.Shape()),
		AllPositive: r.AllPositive(),
		NarrowKeys:  r.Keys() == compact.Keys4,
		Fields:      fields,
		Message:     base64.StdEncoding.EncodeToString(data),

		RustByteExact: everyStringIsRaw(value),
	}
	for range r.Records() {
		c.Records = append(c.Records, readCompositeRecord(name, r, fields))
	}
	if err := r.Err(); err != nil {
		panic(fmt.Sprintf("%s: %v", name, err))
	}
	// The corpus is only as good as its round trip.
	dst := reflect.New(reflect.TypeOf(value))
	if err := colbin.Unmarshal(data, dst.Interface()); err != nil {
		panic(fmt.Sprintf("%s: round trip: %v", name, err))
	}
	return c
}

// --- the schema tree ---------------------------------------------------------

// compositeSchema is main.go's fieldsOf with a recursive kind. The ids still come
// from a schema section colbin wrote, one MarshalJSON per struct type, so nothing
// here re-derives the id hash.
func compositeSchema(t reflect.Type) []schemaOut {
	flat := fieldIDsOf(t)
	sfs := encodableFields(t)
	out := make([]schemaOut, len(flat))
	for i := range flat {
		out[i] = schemaOut{Name: flat[i].Name, ID: flat[i].ID, Kind: kindNodeOf(sfs[i].Type)}
	}
	return out
}

func kindNodeOf(t reflect.Type) kindNode {
	switch t.Kind() {
	case reflect.Struct:
		return kindNode{Kind: "struct", Fields: compositeSchema(t)}
	case reflect.Map:
		key, val := kindNodeOf(t.Key()), kindNodeOf(t.Elem())
		return kindNode{Kind: "map", Key: &key, Val: &val}
	case reflect.Slice:
		if t.Elem().Kind() == reflect.Uint8 {
			return kindNode{Kind: "bytes"}
		}
		// A slice of scalars keeps its bulk-codec name; anything else is an
		// element-by-element array, which is the distinction the wire makes.
		if isScalarKind(t.Elem()) {
			return kindNode{Kind: kindName(t)}
		}
		elem := kindNodeOf(t.Elem())
		return kindNode{Kind: "array", Elem: &elem}
	}
	return kindNode{Kind: kindName(t)}
}

// isScalarKind reports whether a slice of t goes to the bulk codecs rather than
// being written element by element -- the same split compactValueOp makes.
func isScalarKind(t reflect.Type) bool {
	switch t.Kind() {
	case reflect.Bool, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
		reflect.Float32, reflect.Float64, reflect.String:
		return true
	}
	return false
}

// --- reading the message back ------------------------------------------------

func readCompositeRecord(name string, r *compact.Reader, fields []schemaOut) []ccField {
	byID := map[int]kindNode{}
	for _, f := range fields {
		byID[f.ID] = f.Kind
	}
	record := []ccField{}
	for {
		key := r.Key()
		if err := r.Err(); err != nil {
			panic(fmt.Sprintf("%s: %v", name, err))
		}
		if key == compact.TerminatorKey {
			return record
		}
		kind, ok := byID[int(key)]
		if !ok {
			panic(fmt.Sprintf("%s: message names field id %d, which the schema does not have", name, key))
		}
		record = append(record, ccField{ID: int(key), V: readCompositeValue(name, r, kind)})
	}
}

// readCompositeValue mirrors what the Rust decoder must do: dispatch on the
// schema, and recurse for the three composite forms.
func readCompositeValue(name string, r *compact.Reader, kind kindNode) any {
	switch kind.Kind {
	case "struct":
		return readCompositeRecord(name, r, kind.Fields)
	case "array":
		n, ok := r.Count(minBitsOf(r, *kind.Elem))
		if !ok {
			panic(fmt.Sprintf("%s: bad array count", name))
		}
		out := []any{}
		for range n {
			out = append(out, readCompositeValue(name, r, *kind.Elem))
		}
		return out
	case "map":
		n, ok := r.Count(minBitsOf(r, *kind.Key) + minBitsOf(r, *kind.Val))
		if !ok {
			panic(fmt.Sprintf("%s: bad map count", name))
		}
		out := []ccEntry{}
		for range n {
			k := readCompositeValue(name, r, *kind.Key)
			v := readCompositeValue(name, r, *kind.Val)
			out = append(out, ccEntry{K: k, V: v})
		}
		return out
	}
	return readCompact(r, kind.Kind)
}

// minBitsOf is compactElemMinBits, which the count bounds check needs.
func minBitsOf(r *compact.Reader, kind kindNode) int {
	switch kind.Kind {
	case "bool":
		return 1
	case "f32":
		return 32
	case "f64":
		return 64
	case "struct":
		return r.Keys().Bits()
	}
	return 8
}

File: rust/tests/composites.rs
//! The composite corpus: compact-mode messages carrying nested structs, arrays
//! of structs, maps and nested slices.
//!
//! `vectors/composites.go` writes it, forcing every case through compact mode
//! with `colbin.MarshalForceCompact` — `Marshal` picks the mode on size for a
//! composite type, and a case that quietly came out columnar would test nothing
//! here. As in the flat corpus, every expected value was read back out of the
//! bytes by the real Go compact reader, so what is pinned is what the format
//! wrote, presence included.
//!
//! The schema is a tree here rather than a string per field, which is why this is
//! a second file and a second parser: the 26 flat vectors keep pinning the format
//! as it stood before composites existed.

use colbin::{Kind, Schema, Value};
use serde_json::Value as Json;

mod common;
use common::{assert_record, decode_base64, kind_of};

const CORPUS: &str = include_str!("../vectors/composites.json");

fn corpus() -> Json {
    serde_json::from_str(CORPUS).expect("the corpus is valid JSON")
}

/// Every composite message, decoded against the schema tree its Go type implies,
/// must yield the values and the presence the Go decoder read out of the same
/// bytes.
#[test]
fn decodes_the_go_composite_corpus() {
    let corpus = corpus();
    let cases = corpus["cases"].as_array().expect("cases is an array");
    assert!(!cases.is_empty());
    for case in cases {
        let name = case["name"].as_str().unwrap();
        let schema = schema_of(&case["fields"]);
        let message = decode_base64(case["message"].as_str().unwrap());

        // Every case is compact by construction; a case that stopped being would
        // stop testing the recursion it was written for.
        assert!(colbin::is_compact(&message), "{name} is not compact");
        // The header facts the Go writer recorded, so a key-width or sign
        // decision that drifted is caught here rather than as a value mismatch.
        assert_eq!(
            schema.narrow_keys(),
            case["narrowKeys"].as_bool().unwrap(),
            "{name}: narrow keys"
        );

        let records =
            colbin::decode(&message, &schema).unwrap_or_else(|err| panic!("{name}: {err}"));
        let want = case["records"].as_array().expect("records is an array");
        assert_eq!(records.len(), want.len(), "{name}: record count");
        for (index, (got, want)) in records.iter().zip(want).enumerate() {
            assert_record(&format!("{name} record {index}"), got, want, &schema);
        }
    }
}

/// A nested struct all of whose fields are zero is omitted entirely, so the
/// record does not name it at all — the omit-zero rule reaching into a composite
/// is worth asserting directly rather than only through a value comparison.
#[test]
fn a_zero_nested_struct_is_absent() {
    let corpus = corpus();
    let case = case_named(&corpus, "cc-nested-zero");
    let schema = schema_of(&case["fields"]);
    let message = decode_base64(case["message"].as_str().unwrap());
    let record = colbin::decode_one(&message, &schema).expect("decodes");

    assert!(record.get(2).is_none(), "the zero nested struct was named");
    assert_eq!(record.i64(1), 7);
    assert_eq!(record.str(3), "outer");
}

/// The three composite values come back as the tree the schema described, so a
/// caller reads them without knowing how they were framed.
#[test]
fn composites_decode_into_the_value_tree() {
    let corpus = corpus();

    let case = case_named(&corpus, "cc-nested-struct");
    let schema = schema_of(&case["fields"]);
    let record =
        colbin::decode_one(&decode_base64(case["message"].as_str().unwrap()), &schema).unwrap();
    let sub = record.get(2).and_then(Value::as_record).expect("a record");
    assert_eq!(sub.i64(1), -3);
    assert_eq!(sub.str(2), "inner");
    assert!(sub.bool(3));

    let case = case_named(&corpus, "cc-rows");
    let schema = schema_of(&case["fields"]);
    let record =
        colbin::decode_one(&decode_base64(case["message"].as_str().unwrap()), &schema).unwrap();
    let rows = record.get(2).and_then(Value::as_array).expect("an array");
    assert_eq!(rows.len(), 3);
    assert_eq!(rows[0].as_record().unwrap().str(2), "a");
    assert_eq!(rows[1].as_record().unwrap().i64(1), -2);
    // The third row held nothing, so its key run is a bare terminator.
    assert!(rows[2].as_record().unwrap().is_empty());

    let case = case_named(&corpus, "cc-maps");
    let schema = schema_of(&case["fields"]);
    let record =
        colbin::decode_one(&decode_base64(case["message"].as_str().unwrap()), &schema).unwrap();
    let entries = record.get(1).and_then(Value::as_map).expect("a map");
    assert_eq!(entries.len(), 1);
    assert_eq!(entries[0].0, Value::String("alpha".into()));
    assert_eq!(entries[0].1, Value::Int(-2));
}

/// A schema nested past the ceiling is refused when it is built, so the readers
/// never descend far enough to exhaust the stack.
#[test]
fn a_schema_nested_too_deep_is_refused() {
    let mut kind = Kind::Int32;
    let mut built = 0;
    loop {
        match Schema::from_ids([(1, kind.clone())]) {
            Ok(schema) => {
                built += 1;
                assert!(
                    built <= colbin::MAX_DEPTH,
                    "built {built} levels, past the ceiling"
                );
                kind = schema.nested();
            }
            Err(err) => {
                assert_eq!(err, colbin::Error::TooDeep(colbin::MAX_DEPTH + 1));
                assert_eq!(built, colbin::MAX_DEPTH);
                return;
            }
        }
    }
}

/// A composite kind in a standard-mode column is refused rather than panicked
/// on: the columnar layout carries these as sub-tables, which this decoder does
/// not read.
#[test]
fn a_composite_in_a_standard_column_is_refused() {
    let inner = Schema::from_ids([(1, Kind::Int32)]).unwrap();
    let schema = Schema::from_ids([(1, Kind::Int64), (2, inner.nested())]).unwrap();
    // A standard-mode message: byte 0 is an even version byte. The record count
    // and the first column are enough to reach field 2's descriptor.
    let message = [0x02, 4, 2, 1, 0x00, 0, 0, 0, 0, 2, 0x05];
    match colbin::decode(&message, &schema) {
        Err(colbin::Error::CompositeUnsupported { id: 2 }) => {}
        // A truncated or type-mismatched column is an equally acceptable refusal;
        // what must not happen is a panic or a silent wrong value.
        Err(_) => {}
        Ok(records) => panic!("a composite standard column decoded: {records:?}"),
    }
}

// --- the schema tree ---------------------------------------------------------

fn case_named<'a>(corpus: &'a Json, name: &str) -> &'a Json {
    corpus["cases"]
        .as_array()
        .unwrap()
        .iter()
        .find(|case| case["name"] == name)
        .unwrap_or_else(|| panic!("the corpus has no case {name}"))
}

/// Builds the schema a case describes. The ids are taken as given rather than
/// rehashed: `field_ids_match_the_go_layout` in the flat corpus is what pins the
/// hash, and every struct here is tagged or deliberately not.
fn schema_of(fields: &Json) -> Schema {
    let specs: Vec<(u8, Kind)> = fields
        .as_array()
        .expect("fields is an array")
        .iter()
        .map(|field| {
            let id = field["id"].as_u64().expect("a field id") as u8;
            (id, kind_tree(&field["k"]))
        })
        .collect();
    Schema::from_ids(specs).expect("the corpus describes a valid schema")
}

/// One kind node, which for the three composite forms holds kinds of its own.
fn kind_tree(node: &Json) -> Kind {
    match node["kind"].as_str().expect("a kind name") {
        "struct" => schema_of(&node["fields"]).nested(),
        "array" => Kind::Array(Box::new(kind_tree(&node["elem"]))),
        "map" => Kind::Map(
            Box::new(kind_tree(&node["key"])),
            Box::new(kind_tree(&node["val"])),
        ),
        scalar => kind_of(scalar),
    }
}
